# Data driven customer churn prediction for SyriaTel company

* ## Business understanding

## Overview

In the telecommunications industry, customer churn refers to the loss of subscribers who stop using a company's service during a specific period. Customer churn is also refered to as customer attrition or subscriber churn. There are two forms of customer churn; voluntary churn and involuntary churn. Voluntary churn involves consumers leaving on their own due to various factors e.g. poor customer service, better options elsewhere etc. On the other hand involuntary churn involves the service provider deliberately disconnecting some subscribers due to one reason or the other. High churn rates typically have adverse effects on telecomm companies, both to their reputation and to their balance sheet. It is therefore my initiative to attempt to use the historical data available and machine learning algorithms so as to be able to predict which subscribers will leave and also the churn rate, within that specified period under review. The insights gain from this will enable us to effective advice management, and help develop a strategic response and a prevetative measure.


## Business problem

High churn rates are one of the major risks that telecomm companies face all over the world. There are multiple causes of customer churn, and SyriaTel company just like any other company in other industries is affected by the 5 forces that can affect a companies profitability. These are what we call the 'Porters 5 forces' and include:
1. Competitive rivalry
2. Threat of substitutes
3. Threat of new market entrants
4. Buyer power
5. Supplier power

Basically, these forces coming from different directions and from different sources but eventually if the company does not take decisive and strategic measures to neutralize them, each one has the same final adverse result, which include reduced profitability, shrinking market share, financial distress among other things. It is therefore imperative, for SyriaTel to develop a strategy to ensure that it reduces the risk of suffering from customer churn, and indeed the most effective solution has to be data-driven. Studies have shown that indeed it costs more to acquire a new customer compared to what it costs just to retain the one's in a company's customer base. Due to such factors and the related risks, it is important to apply Machine learning so as to proactively determine the customers who are at risk, and in turn take the necessary measure to mitigate that risk.


* ##  Data understanding

The dataset being used is sourced from Kaggle, specifically the "SyriaTel Customer Churn" dataset. It can be found at the following link:
https://www.kaggle.com/datasets/becksddf/churn-in-telecoms-dataset/data

For this section of the project, we will be using the "SyriaTel Customer Churn" dataset from Kaggle. The dataset contains various features related to customer demographics, service usage, and churn status. It is a valuable resource for analyzing customer behavior and predicting churn in the telecommunications industry. I will rely on this dataset to build a predictive model that can help identify customers at risk of churn and suggest strategies to retain them.

In [51]:
# Importing the necessary libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.linear_model import LogisticRegression
from sklearn.decomposition import PCA
from sklearn.linear_model import LinearRegression
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import classification_report, roc_auc_score, roc_curve
from imblearn.over_sampling import SMOTE
from sklearn.feature_selection import RFE

1. Loading the dataset, and converting it to a pandas DataFrame.

In [52]:
# Loading the dataset
churn_data = pd.read_csv('syriatel_customerchurn.csv')

2. To check the general structure of the dataset and its metadata.

In [53]:
# Display the structure of the dataset

churn_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3333 entries, 0 to 3332
Data columns (total 21 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   state                   3333 non-null   object 
 1   account length          3333 non-null   int64  
 2   area code               3333 non-null   int64  
 3   phone number            3333 non-null   object 
 4   international plan      3333 non-null   object 
 5   voice mail plan         3333 non-null   object 
 6   number vmail messages   3333 non-null   int64  
 7   total day minutes       3333 non-null   float64
 8   total day calls         3333 non-null   int64  
 9   total day charge        3333 non-null   float64
 10  total eve minutes       3333 non-null   float64
 11  total eve calls         3333 non-null   int64  
 12  total eve charge        3333 non-null   float64
 13  total night minutes     3333 non-null   float64
 14  total night calls       3333 non-null   

The dataset contains 3333 rows and 21 columns, with the target variable 'Churn' being categorical.

Float datatypes = 8

Integer datatypes = 8

Object datatypes = 4

Boolean datatypes = 1

In [54]:
# Display the first 5 rows of the dataset

churn_data.head()

state  account length  area code phone number international plan  \
0    KS             128        415     382-4657                 no   
1    OH             107        415     371-7191                 no   
2    NJ             137        415     358-1921                 no   
3    OH              84        408     375-9999                yes   
4    OK              75        415     330-6626                yes   

  voice mail plan  number vmail messages  total day minutes  total day calls  \
0             yes                     25              265.1              110   
1             yes                     26              161.6              123   
2              no                      0              243.4              114   
3              no                      0              299.4               71   
4              no                      0              166.7              113   

   total day charge  ...  total eve calls  total eve charge  \
0             45.07  ...               99             16.78   
1             27.47  ...              103             16.62   
2             41.38  ...              110             10.30   
3             50.90  ...               88              5.26   
4             28.34  ...              122             12.61   

   total night minutes  total night calls  total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   total intl minutes  total intl calls  total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   customer service calls  churn  
0                       1  False  
1                       1  False  
2                       0  False  
3                       2  False  
4                       3  False  

[5 rows x 21 columns]

3. Next we check the summary statistics of the colums with numerical values, and the unique values of the categorical columns.

In [55]:
# Summary statistics of numerical columns

churn_data.describe()

account length    area code  number vmail messages  total day minutes  \
count     3333.000000  3333.000000            3333.000000        3333.000000   
mean       101.064806   437.182418               8.099010         179.775098   
std         39.822106    42.371290              13.688365          54.467389   
min          1.000000   408.000000               0.000000           0.000000   
25%         74.000000   408.000000               0.000000         143.700000   
50%        101.000000   415.000000               0.000000         179.400000   
75%        127.000000   510.000000              20.000000         216.400000   
max        243.000000   510.000000              51.000000         350.800000   

       total day calls  total day charge  total eve minutes  total eve calls  \
count      3333.000000       3333.000000        3333.000000      3333.000000   
mean        100.435644         30.562307         200.980348       100.114311   
std          20.069084          9.259435          50.713844        19.922625   
min           0.000000          0.000000           0.000000         0.000000   
25%          87.000000         24.430000         166.600000        87.000000   
50%         101.000000         30.500000         201.400000       100.000000   
75%         114.000000         36.790000         235.300000       114.000000   
max         165.000000         59.640000         363.700000       170.000000   

       total eve charge  total night minutes  total night calls  \
count       3333.000000          3333.000000        3333.000000   
mean          17.083540           200.872037         100.107711   
std            4.310668            50.573847          19.568609   
min            0.000000            23.200000          33.000000   
25%           14.160000           167.000000          87.000000   
50%           17.120000           201.200000         100.000000   
75%           20.000000           235.300000         113.000000   
max           30.910000           395.000000         175.000000   

       total night charge  total intl minutes  total intl calls  \
count         3333.000000         3333.000000       3333.000000   
mean             9.039325           10.237294          4.479448   
std              2.275873            2.791840          2.461214   
min              1.040000            0.000000          0.000000   
25%              7.520000            8.500000          3.000000   
50%              9.050000           10.300000          4.000000   
75%             10.590000           12.100000          6.000000   
max             17.770000           20.000000         20.000000   

       total intl charge  customer service calls  
count        3333.000000             3333.000000  
mean            2.764581                1.562856  
std             0.753773                1.315491  
min             0.000000                0.000000  
25%             2.300000                1.000000  
50%             2.780000                1.000000  
75%             3.270000                2.000000  
max             5.400000                9.000000

This enables us to inspect the details of the values within each column such as the standard deviation, the maximum and minimum values, the mean, and the total count of the non_null values in each column.

It is impportant to note that there are no missing values in any other columns within our dataset, this is quite a relief as we will not have to perform any extreme imputations that might affects integrity of our data.

* ## Data preparation

## Data cleaning

The process of data cleaning involves utilizing data science techniques for removing unnecessary and irrelevant columns, handling missing data through effective imputation, fixing inconsistent labels, removing duplicates etc.

Removing the unnecessary columns i.e. irrevelant data.

In [56]:
# Removing unnecessary columns
churn_data.drop(columns=['phone number', 'state', 'area code'], inplace=True)

In [57]:
churn_data.head()

account length international plan voice mail plan  number vmail messages  \
0             128                 no             yes                     25   
1             107                 no             yes                     26   
2             137                 no              no                      0   
3              84                yes              no                      0   
4              75                yes              no                      0   

   total day minutes  total day calls  total day charge  total eve minutes  \
0              265.1              110             45.07              197.4   
1              161.6              123             27.47              195.5   
2              243.4              114             41.38              121.2   
3              299.4               71             50.90               61.9   
4              166.7              113             28.34              148.3   

   total eve calls  total eve charge  total night minutes  total night calls  \
0               99             16.78                244.7                 91   
1              103             16.62                254.4                103   
2              110             10.30                162.6                104   
3               88              5.26                196.9                 89   
4              122             12.61                186.9                121   

   total night charge  total intl minutes  total intl calls  \
0               11.01                10.0                 3   
1               11.45                13.7                 3   
2                7.32                12.2                 5   
3                8.86                 6.6                 7   
4                8.41                10.1                 3   

   total intl charge  customer service calls  churn  
0               2.70                       1  False  
1               3.70                       1  False  
2               3.29                       0  False  
3               1.78                       2  False  
4               2.73                       3  False

In [58]:
churn_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3333 entries, 0 to 3332
Data columns (total 18 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   account length          3333 non-null   int64  
 1   international plan      3333 non-null   object 
 2   voice mail plan         3333 non-null   object 
 3   number vmail messages   3333 non-null   int64  
 4   total day minutes       3333 non-null   float64
 5   total day calls         3333 non-null   int64  
 6   total day charge        3333 non-null   float64
 7   total eve minutes       3333 non-null   float64
 8   total eve calls         3333 non-null   int64  
 9   total eve charge        3333 non-null   float64
 10  total night minutes     3333 non-null   float64
 11  total night calls       3333 non-null   int64  
 12  total night charge      3333 non-null   float64
 13  total intl minutes      3333 non-null   float64
 14  total intl calls        3333 non-null   

Check whether there are any duplicate values, within our dataset.

In [59]:
# Checking for duplicate Values

churn_data.duplicated().sum()

0

There are no duplicate values within our dataset, as the output of our analysis above shows that there are 0 true duplicate rows within our dataset.

In [60]:
# Our next step is to deal with the categorical variables in our dataset. We will convert them into numerical values using Label Encoding, which is a common technique for handling categorical data in machine learning.
# Converting categorical variables to numerical values
categorical_cols = churn_data.select_dtypes(include=['object']).columns
for col in categorical_cols:
    le = LabelEncoder()
    churn_data[col] = le.fit_transform(churn_data[col])

# Display the first 5 rows after encoding categorical variables
churn_data.head()



account length  international plan  voice mail plan  number vmail messages  \
0             128                   0                1                     25   
1             107                   0                1                     26   
2             137                   0                0                      0   
3              84                   1                0                      0   
4              75                   1                0                      0   

   total day minutes  total day calls  total day charge  total eve minutes  \
0              265.1              110             45.07              197.4   
1              161.6              123             27.47              195.5   
2              243.4              114             41.38              121.2   
3              299.4               71             50.90               61.9   
4              166.7              113             28.34              148.3   

   total eve calls  total eve charge  total night minutes  total night calls  \
0               99             16.78                244.7                 91   
1              103             16.62                254.4                103   
2              110             10.30                162.6                104   
3               88              5.26                196.9                 89   
4              122             12.61                186.9                121   

   total night charge  total intl minutes  total intl calls  \
0               11.01                10.0                 3   
1               11.45                13.7                 3   
2                7.32                12.2                 5   
3                8.86                 6.6                 7   
4                8.41                10.1                 3   

   total intl charge  customer service calls  churn  
0               2.70                       1  False  
1               3.70                       1  False  
2               3.29                       0  False  
3               1.78                       2  False  
4               2.73                       3  False

In [61]:
# Convert boolean to integer
churn_data['churn'] = churn_data['churn'].astype(int)

In [62]:
churn_data.head()

account length  international plan  voice mail plan  number vmail messages  \
0             128                   0                1                     25   
1             107                   0                1                     26   
2             137                   0                0                      0   
3              84                   1                0                      0   
4              75                   1                0                      0   

   total day minutes  total day calls  total day charge  total eve minutes  \
0              265.1              110             45.07              197.4   
1              161.6              123             27.47              195.5   
2              243.4              114             41.38              121.2   
3              299.4               71             50.90               61.9   
4              166.7              113             28.34              148.3   

   total eve calls  total eve charge  total night minutes  total night calls  \
0               99             16.78                244.7                 91   
1              103             16.62                254.4                103   
2              110             10.30                162.6                104   
3               88              5.26                196.9                 89   
4              122             12.61                186.9                121   

   total night charge  total intl minutes  total intl calls  \
0               11.01                10.0                 3   
1               11.45                13.7                 3   
2                7.32                12.2                 5   
3                8.86                 6.6                 7   
4                8.41                10.1                 3   

   total intl charge  customer service calls  churn  
0               2.70                       1      0  
1               3.70                       1      0  
2               3.29                       0      0  
3               1.78                       2      0  
4               2.73                       3      0

* ## Modeling and evaluation

* ## Conclusion